Augustocollado/tanssi templates (#7)
moondance-labs/tanssi-docs: learn/tanssi/templates.md

---

> title: Templates Included in Tanssi

Reviewer: If I think about what templates means I have no clue.

So would change "Templates" for Appchain Templates or something like that

description: Tanssi protocol provides useful templates to start building your Appchain, including a ready-to-use EVM template for Ethereum compatibility.
---

# Templates Included in Tanssi {: #templates-included-in-tanssi } 

## Introduction {: #introduction } 

> Initialize a new repository and start building a new ContainerChain from scratch can be a daunting prospect, but, fortunately, Tanssi includes templates that help developers to jumpstart the process and benefit in some aspects, such as:

Reviewer: ```suggestion
Building a new ContainerChain from scratch can be a daunting prospect. Fortunately, thanks to the [Appchain development framework](/learn/framework/overview){target=_blank} used by Tanssi,  developers can leverage some pre-bundled templates that helps them jumpstart the process and benefit in some aspects, such as:
```

Reviewer: Maybe here talk a bit more about Substrate's modularity without repeating ourselves too much. Just like 1 extra sentence

> - **Head start** - templates provide a starting point for your project, saving significant time and effort by providing a basic structure and a set of tested and ready-to-use functionalities. It allows developers to accelerate the construction of prototypes or minimum viable products (MVPs) and reduce the time to market

Reviewer: I would change "templates" for "Appchain templates" everywhere tbh

- **Consistency** - included templates follow established design patterns, coding standards, and best practices widely accepted among the Substrate developers community. They also provide a default set of architecture definitions to streamline the blockchain development

- **UX** - templates cover the most demanded use cases, such as the EVM support for an Ethereum-compatible ContainerChain

- **Customizability** - templates are a great starting point and are completely customizable. The functionalities and default configurations they include can be modified, replaced, or extended to meet the specific requirements of the use case

- **Upgrade and Compatibility** - Substrate is an evolving framework, with new features, enhancements, and bug fixes being regularly introduced, and the provided Tanssi templates are kept up to date with these upgrades

## Start building a ContainerChain {: #start-building } 

To start building a ContainerChain to deploy in Tanssi, some useful templates to kick-start the development process are provided in the [official repository](https://github.com/moondance-labs/tanssi){target=blank}:

> - **Simple Template** - this node template is based on the [parachain Substrate template](https://github.com/substrate-developer-hub/substrate-parachain-template){target=_blank} featuring [Cumulus](https://github.com/paritytech/cumulus/){target=blank} to allow the ContainerChain to work properly in the Polkadot ecosystem. It also includes Tanssi's own pallets to support the protocol and some other basic pallets such as *pallet-balances* for currency operations

Reviewer: I would create a `###` section for each template so you can expand a bit what each thing is. You have to consider that someone coming from outside has no clue what Cumulus. So you can go a bit into more detail about the core pallets included and why would you need them

> - **Ethereum Virtual Machine** - this node template is based on [Frontier](https://github.com/paritytech/frontier){target=_blank}, a project that adds an Ethereum compatibility layer for Substrate. This template provides not only Tanssi protocol support but an EVM and full Ethereum support (exposing the expected RPC endpoints and emulating Ethereum's block production mechanism) allowing existing DApps to be deployed and run smoothly on your Substrate Appchain and seamlessly integrate and use popular tools such as [MetaMask](https://metamask.io){target=blank}, [Truffle](https://trufflesuite.com){target=blank}, [Remix](https://remix.ethereum.org){target=blank}, block explorers and many other

Reviewer: As before, add a `###` for this section.

Here you should use keywords like Smart Contracts, and that you can use tools like Metamask, Hardhat, Remix, Foundry (Truffle is not really famous anymore) out of the box, with no extra adapters

- **Consistency** - included Appchain templates follow established design patterns, coding standards, and best practices widely accepted among the Substrate developers community. They also provide a default set of architecture definitions to streamline the blockchain development

- **UX** - Appchain templates cover the most demanded use cases, such as the EVM support for an Ethereum-compatible ContainerChain

- **Customizability** - Appchain templates are a great starting point and are completely customizable. The functionalities and default configurations they include can be modified, replaced, or extended to meet the specific requirements of the use case

- **Upgrade and Compatibility** - Substrate is an evolving framework, with new features, enhancements, and bug fixes being regularly introduced, and the provided Tanssi Appchain templates are kept up to date with these upgrades

> ## Start building a ContainerChain {: #start-building } 

Reviewer 2: ```suggestion
## Start Building a ContainerChain {: #start-building } 
```

To start building a ContainerChain to deploy in Tanssi, some useful Appchain templates to kick-start the development process are provided in the [official repository](https://github.com/moondance-labs/tanssi){target=blank}.

The process is as simple as:

1. Select one of the templates

> 2. Add the specific logic to adapt the runtime to the requeriments of the use case

Reviewer 2: ```suggestion
2. Add the specific logic to adapt the runtime to the requirements of the use case
```

3. Deploy in Tanssi

![Using Templates to Speed Up the Developmet Process](/images/learn/tanssi/templates/templates-1.png)

The two included templates are *Simple* and *EVM*, which are presented in the following sections.

### Simple Template  {: #simple-template } 

> As presented in the [Overview](/learn/tanssi/overview) article, Appchains deployed through Tanssi (ContainerChains) are fully sovereign customizable Layer 1 solutions that leverage the benefits of becoming part of the wider Polkadot ecosystem.

Reviewer 2: ```suggestion
As presented in the [Overview](/learn/tanssi/overview) article, Appchains deployed through Tanssi (ContainerChains) are fully sovereign, customizable Layer 1 solutions that leverage the benefits of becoming part of the wider Polkadot ecosystem.
```

As a sibling chain of the other members of the ecosystem (like Tanssi network itself), they must include the necessary behavior to implement their consensus mechanism and be able to interact and synchronize with Polkadot's relay chain.

> Simple Appchain template is based on the [parachain Substrate template](https://github.com/substrate-developer-hub/substrate-parachain-template){target=_blank} featuring [Cumulus](https://github.com/paritytech/cumulus/){target=blank} SDK, which includes all the required functionality for the collators logic, p2p, database, and synchronization layers between the ContainerChain and the relay chain, making the integration of a new ContainerChain into the Polkadot ecosystem a breeze.

Reviewer 2: ```suggestion
The Simple Appchain template is based on the [Substrate Parachain Template
](https://github.com/substrate-developer-hub/substrate-parachain-template){target=_blank} featuring [Cumulus](https://github.com/paritytech/cumulus/){target=blank} SDK, which includes all the required functionality for the collators logic, p2p, database, and synchronization layers between the ContainerChain and the relay chain, making the integration of a new ContainerChain into the Polkadot ecosystem a breeze.
```

> This template also includes Tanssi's pallet [authorities_noting](https://github.com/moondance-labs/tanssi/blob/master/container-chains/pallets/authorities-noting/src/lib.rs){target=blank} to support the protocol, implementing the necessary logic to retrieve and validate the collators set assigned to provide block production services to the ContainerChain, and the logic for the collator to sign the block when the consensus mechanism determines that it is its turn to build it (and thus being rewarded accordingly).

Author: Much better!

### Ethereum Virtual Machine Template {: #evm-template } 

Extending the *Simple template*, this Appchain template provides not only Tanssi protocol support but an EVM and full Ethereum compatibility.

Leveraging the [Frontier project](https://github.com/paritytech/frontier){target=_blank}, this template includes an Ethereum compatibility layer for ContainerChains to allow running unmodified Ethereum DApps. 

> Using this template, Containerchains support to deploy and run any existing Smart Contract written in Solidity or Vyper languages with no changes, and by emulating Ethereum block production and exposing the expected RPC interface, developers can also continue using the same tools like [Metamask](https://metamask.io){target=blank}, [Hardhat](https://hardhat.org){target=blank}, [Remix](https://remix.ethereum.org){target=blank}, [Foundry](https://github.com/foundry-rs/foundry){target=blank}, [Truffle](https://trufflesuite.com){target=blank}, and many more out of the box, with no extra adapters.

Reviewer 2: ```suggestion
Using this template, Containerchains support the deployment and running of any existing Smart Contract written in Solidity or Vyper with no changes. By emulating Ethereum block production and exposing the expected RPC interface, developers can also continue using the same tools like [Metamask](https://metamask.io){target=_blank}, [Hardhat](https://hardhat.org){target=_blank}, [Remix](https://remix.ethereum.org){target=_blank}, [Foundry](https://github.com/foundry-rs/foundry){target=_blank}, [Truffle](https://trufflesuite.com){target=_blank}, and many more out of the box, with no extra adapters.
```

> With this EVM template, developers can have a [Moonbeam](https://moonbeam.network){target=blank}-like ContainerChain in no time and add their custom logic and features specific to the use case.

Reviewer 2: ```suggestion
With this EVM template, developers can have a [Moonbeam](https://moonbeam.network){target=blank}-like ContainerChain in no time and add their own custom logic and features specific to their own use case.
```

- **Consistency** - included Appchain templates follow established design patterns, coding standards, and best practices widely accepted among the Substrate developer community. They also provide a default set of architecture definitions to streamline blockchain development

- **UX** - Appchain templates cover the most demanded use cases, such as the EVM support for an Ethereum-compatible ContainerChain

- **Customizability** - Appchain templates are a great starting point and are completely customizable. The functionalities and default configurations they include can be modified, replaced, or extended to meet the specific requirements of the use case

- **Upgrades and Compatibility** - Substrate is an evolving framework, with new features, enhancements, and bug fixes being regularly introduced. The provided Tanssi Appchain templates are kept up-to-date with these upgrades

## Start Building a ContainerChain {: #start-building } 

> To start building a ContainerChain to deploy in Tanssi, some useful Appchain templates to kick-start the development process are provided in the [official repository](https://github.com/moondance-labs/tanssi){target=blank}.

Reviewer 2: ```suggestion
To start building a ContainerChain to deploy in Tanssi, some useful Appchain templates to kick-start the development process are provided in the [official repository](https://github.com/moondance-labs/tanssi){target=_blank}.
```

The process is as simple as:

1. Select one of the templates
2. Add the specific logic to adapt the runtime to the requirements of the use case
3. Deploy in Tanssi

![Using Templates to Speed Up the Developmet Process](/images/learn/tanssi/templates/templates-1.png)

The two included templates are *Baseline Appchain Template* and *Baseline EVM Template*, which are presented in the following sections.

### Baseline Appchain Template  {: #baseline-appchain-template } 

As presented in the [Overview](/learn/tanssi/overview) article, Appchains deployed through Tanssi (ContainerChains) are fully sovereign, customizable Layer 1 solutions that leverage the benefits of becoming part of the wider Polkadot ecosystem.

As a sibling chain of the other members of the ecosystem (like the Tanssi network itself), they must include the necessary behavior to implement their consensus mechanism and be able to interact and synchronize with Polkadot's relay chain.

> The Baseline Appchain Template is based on the [Substrate Parachain Template
> ](https://github.com/substrate-developer-hub/substrate-parachain-template){target=_blank} featuring [Cumulus](https://github.com/paritytech/cumulus/){target=blank} SDK, which includes all the required functionality for the collators logic, p2p, database, and synchronization layers between the ContainerChain and the relay chain, making the integration of a new ContainerChain into the Polkadot ecosystem a breeze.

Reviewer 2: ```suggestion
The Baseline Appchain Template is based on the [Substrate Parachain Template](https://github.com/substrate-developer-hub/substrate-parachain-template){target=_blank} featuring [Cumulus](https://github.com/paritytech/cumulus/){target=_blank} SDK, which includes all the required functionality for the collators logic, p2p, database, and synchronization layers between the ContainerChain and the relay chain, making the integration of a new ContainerChain into the Polkadot ecosystem a breeze.
```

> This template also includes Tanssi's [Authorities Noting](https://github.com/moondance-labs/tanssi/blob/master/container-chains/pallets/authorities-noting/src/lib.rs){target=blank} pallet, which implements the logic for retrieving and validating the set of collators assigned to provide block production services to the ContainerChain. It also includes logic that allows a collator to sign the block when the consensus mechanism determines that it is the collator's turn to produce the block (and thus be rewarded accordingly).

Reviewer 2: ```suggestion
This template also includes Tanssi's [Authorities Noting](https://github.com/moondance-labs/tanssi/blob/master/container-chains/pallets/authorities-noting/src/lib.rs){target=_blank} pallet, which implements the logic for retrieving and validating the set of collators assigned to provide block production services to the ContainerChain. It also includes logic that allows a collator to sign the block when the consensus mechanism determines that it is the collator's turn to produce the block (and thus be rewarded accordingly).
```

### Baseline EVM (Ethereum Virtual Machine) Template {: #baseline-evm-template } 

Extending the *Baseline Appchain template*, this template provides not only Tanssi protocol support but an EVM and full Ethereum compatibility.

Leveraging the [Frontier project](https://github.com/paritytech/frontier){target=_blank}, this template includes an Ethereum compatibility layer for ContainerChains to allow running unmodified Ethereum DApps. 

> Using this template, Containerchains support the deployment and running of any existing Smart Contract written in Solidity or Vyper with no changes. By emulating Ethereum block production and exposing the expected RPC interface, developers can also continue using the same tools like [Metamask](https://metamask.io){target=_blank}, [Hardhat](https://hardhat.org){target=_blank}, [Remix](https://remix.ethereum.org){target=_blank}, [Foundry](https://github.com/foundry-rs/foundry){target=_blank}, [Truffle](https://trufflesuite.com){target=_blank}, and many more out of the box, with no extra adapters.

Reviewer 2: ```suggestion
Using this template, ContainerChains support the deployment and running of any existing Smart Contract written in Solidity or Vyper with no changes. By emulating Ethereum block production and exposing the expected RPC interface, developers can also continue using the same tools like [Metamask](https://metamask.io){target=_blank}, [Hardhat](https://hardhat.org){target=_blank}, [Remix](https://remix.ethereum.org){target=_blank}, [Foundry](https://github.com/foundry-rs/foundry){target=_blank}, [Truffle](https://trufflesuite.com){target=_blank}, and many more out of the box, with no extra adapters.
```

> With this EVM template, developers can have a [Moonbeam](https://moonbeam.network){target=blank}-like ContainerChain in no time and add their own custom logic and features specific to their own use case.

Reviewer 2: ```suggestion
With this EVM template, developers can have a [Moonbeam](https://moonbeam.network){target=_blank}-like ContainerChain in no time and add their own custom logic and features specific to their own use case.
```

---
title: Appchain Templates Included in Tanssi
description: Tanssi protocol provides useful templates to start building your Appchain, including a ready-to-use EVM template for Ethereum compatibility.
---

# Appchain Templates Included in Tanssi {: #appchain-templates-included-in-tanssi } 

## Introduction {: #introduction } 

> Building a new ContainerChain from scratch can be a daunting prospect. Fortunately, thanks to the [Appchain development framework](/learn/framework/overview){target=_blank} used by Tanssi and its modular-oriented architecture, developers can leverage some pre-bundled Appchain templates that help them jumpstart the process and benefit in some aspects, such as:

Reviewer: Now I'm confused as to if we are using ContainerChains or Appchains generally 😬

- **Customizability** - Appchain templates are a great starting point and are completely customizable. The functionalities and default configurations they include can be modified, replaced, or extended to meet the specific requirements of the use case

- **Upgrade and Compatibility** - Substrate is an evolving framework, with new features, enhancements, and bug fixes being regularly introduced, and the provided Tanssi Appchain templates are kept up to date with these upgrades

## Start building a ContainerChain {: #start-building } 

To start building a ContainerChain to deploy in Tanssi, some useful Appchain templates to kick-start the development process are provided in the [official repository](https://github.com/moondance-labs/tanssi){target=blank}.

The process is as simple as:

1. Select one of the templates
2. Add the specific logic to adapt the runtime to the requeriments of the use case
3. Deploy in Tanssi

![Using Templates to Speed Up the Developmet Process](/images/learn/tanssi/templates/templates-1.png)

The two included templates are *Simple* and *EVM*, which are presented in the following sections.

> ### Simple Template  {: #simple-template } 

Reviewer: I don't really like the name "Simple Template"

Maybe we can do "Baseline Substrate Template" and "Baseline EVM Template"?

Reviewer 2: Maybe consider putting the templates in their own section with whatever names you decide on for the templates:

- Start Building a ContainerChain
- Available Templates
    - ~~Simple~~ ? Template
    - EVM Template

Author: Let's go with "Baseline Appchain Template"
